Sources/protoc-gen-swiftgrpc/Generator-Server.swift

```diff
@@ -63,44 +82,84 @@ extension Generator {
+      println("/// Returns nil for methods not handled by this service.")
+      println("\(access) func handleMethod(_ methodName: String, request: HTTPRequestHead, serverHandler: GRPCChannelHandler, channel: Channel) -> GRPCCallHandler? {")
       indent()
-      switch streamingType(method) {
-      case .unary, .serverStreaming:
-        println("return try \(methodSessionName)Base(")
+      println("switch methodName {")
+      for method in service.methods {
+        self.method = method
+        println("case \"\(method.name)\":")
         indent()
-        println("handler: handler,")
-        println("providerBlock: { try self.\(methodFunctionName)(request: $0, session: $1 as! \(methodSessionName)Base) })")
+        let callHandlerType: String
+        switch streamingType(method) {
+        case .unary: callHandlerType = "UnaryCallHandler"
+        case .serverStreaming: callHandlerType = "ServerStreamingCallHandler"
+        case .clientStreaming: callHandlerType = "ClientStreamingCallHandler"
+        case .bidirectionalStreaming: callHandlerType = "BidirectionalStreamingCallHandler"
+        }
+        println("return \(callHandlerType)(channel: channel, request: request) { context in")
         indent()
-        println(".run()")
+        switch streamingType(method) {
+        case .unary, .serverStreaming:
+          println("return { request in")
+          indent()
+          println("self.\(methodFunctionName)(request: request, context: context)")
+          outdent()
+          println("}")
+        case .clientStreaming, .bidirectionalStreaming:
+          println("return self.\(methodFunctionName)(context: context)")
+        }
         outdent()
+        println("}")
         outdent()
-      default:
-        println("return try \(methodSessionName)Base(")
-        indent()
-        println("handler: handler,")
-        println("providerBlock: { try self.\(methodFunctionName)(session: $0 as! \(methodSessionName)Base) })")
+        println()
+      }
+      println("default: return nil")
+      println("}")
+      outdent()
+      println("}")
+    } else {
+      println("/// Determines and calls the appropriate request handler, depending on the request's method.")
+      println("/// Throws `HandleMethodError.unknownMethod` for methods not handled by this service.")
+      println("\(access) func handleMethod(_ method: String, handler: Handler) throws -> ServerStatus? {")
+      indent()
+      println("switch method {")
+      for method in service.methods {
+        self.method = method
+        println("case \(methodPath):")
         indent()
-        println(".run()")
-        outdent()
+        switch streamingType(method) {
+        case .unary, .serverStreaming:
+          println("return try \(methodSessionName)Base(")
+          indent()
+          println("handler: handler,")
+          println("providerBlock: { try self.\(methodFunctionName)(request: $0, session: $1 as! \(methodSessionName)Base) })")
```

Comment:
Probably shouldn't be force casting like this in generated code

Reply:
This is actually the old, non-NIO implementation. The force cast is necessary but safe in this case (note that we are instantiating `\(methodSessionName)Base(` above).